src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/If.cs

```diff
@@ -83,49 +82,94 @@ public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DTyp
+            // For PowerFxV1 compat rules, validate that all non-predicate args can be coerced to the first one
             for (var i = 1; i < count;)
             {
                 var nodeArg = args[i];
                 var typeArg = argTypes[i];
 
-                var typeSuper = DType.Supertype(
-                    type, 
-                    typeArg,
-                    useLegacyDateTimeAccepts: false,
-                    usePowerFxV1CompatibilityRules: context.Features.PowerFxV1CompatibilityRules);
-                if (!typeSuper.IsError)
+                if (context.Features.PowerFxV1CompatibilityRules)
                 {
-                    type = typeSuper;
-                }
-                else if (typeArg.IsVoid)
-                {
-                    type = DType.Void;
-                }
-                else if (typeArg.IsError)
-                {
-                    errors.EnsureError(args[i], TexlStrings.ErrTypeError);
-                    fArgsValid = false;
-                }
-                else if (!type.IsError)
-                {
-                    if (typeArg.CoercesTo(type, aggregateCoercion: true, isTopLevelCoercion: false, usePowerFxV1CompatibilityRules: context.Features.PowerFxV1CompatibilityRules))
+                    if (typeArg.IsVoid)
+                    {
+                        type = DType.Void;
+                    }
+                    else if (typeArg.IsError)
+                    {
+                        errors.EnsureError(args[i], TexlStrings.ErrTypeError);
+                        fArgsValid = false;
+                    }
+                    else if (type.Kind == DKind.ObjNull)
+                    {
+                        // Anything goes with null
+                        type = typeArg;
+                    }
+                    else if (typeArg.Kind == DKind.ObjNull)
                     {
-                        CollectionUtils.Add(ref nodeToCoercedTypeMap, nodeArg, type);
+                        // ObjNull can be accepted by the current type
+                    }
+                    else if (DType.TryUnionWithCoerce(
+                             type,
+                             typeArg,
+                             usePowerFxV1CompatibilityRules: true,
+                             coerceToLeftTypeOnly: true,
+                             out var unionType,
+                             out var coercionNeeded))
+                    {
+                        type = unionType;
+                        if (coercionNeeded)
+                        {
+                            CollectionUtils.Add(ref nodeToCoercedTypeMap, nodeArg, type);
+                        }
                     }
                     else
                     {
-                        // If the types are incompatible, the result type is void.
+                        // If types are incompatible, result is Void
                         type = DType.Void;
                     }
                 }
```

Comment:
Is this logic shareable with Switch/Coalesce/IfError/..? Should/could it be extracted to a helper?

Reply:
Yes, next step will be to do the same for Switch/IfError. At that time we'll refactor it.